File: 125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start
class Solution:

  def isPalindrome(self, s: str) -> bool:
    if not s:
      return True

    p1 = 0
    p2 = len(s) - 1

    while p1 <= p2:
      if not s[p1].isalnum():
        p1 += 1
        continue

      if not s[p2].isalnum():
        p2 -= 1
        continue

      if s[p1].lower() == s[p2].lower():
        p1 += 1
        p2 -= 1
        continue

      return False

    return True
# @lc code=end


if __name__ == "__main__":
  tests = [
      ("0P", False),
      ("aa", True),
      ("A man, a plan, a canal: Panama", True),
      ("race a car", False),
      (" ", True),
      ("", True),
      ("1", True),
  ]

  s = Solution()
  for test, expected in tests:
    result = s.isPalindrome(test)
    assert result == expected, f"{test}: expected {expected}, got {result}"

File: 125.valid-palindrome.cpp
#include <iostream>
#include <string>
#include <cassert>
#include <vector>
#include <utility>
using namespace std;

/*
 * @lc app=leetcode id=125 lang=cpp
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution
{
public:
  bool isPalindrome(string s)
  {
    char *p1 = &s[0];
    char *p2 = &s[s.length() - 1];

    while (p1 < p2)
    {
      if (!isalnum(*p1))
      {
        p1++;
        continue;
      }

      if (!isalnum(*p2))
      {
        p2--;
        continue;
      }

      if (tolower(*p1) == tolower(*p2))
      {
        p1++;
        p2--;
        continue;
      }

      return false;
    }
    return true;
  }
};
// @lc code=end

int main()
{
  vector<pair<string, bool>> tests = {
    {"A man, a plan, a canal: Panama", true},
    {"race a car", false}
  };

  Solution s;

  for (auto test : tests)
  {
    bool expected = test.second;
    bool result = s.isPalindrome(test.first);
    assert(expected == result);
  }
}
